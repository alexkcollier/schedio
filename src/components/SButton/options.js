export {
  colors,
  allShadeOptions,
  colorNames,
} from '@/utils/component-options'

export const sizes = [
  'small',
  'regular',
  'large',
]

export const types = [
  'outlined',
  'text',
]
