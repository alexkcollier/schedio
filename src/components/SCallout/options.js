export const calloutTypes = ['', 'success', 'warning', 'error']
