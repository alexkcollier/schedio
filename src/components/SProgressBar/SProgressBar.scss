.progress-bar {
  margin-top: spacing();

  &:not(:last-child) {
    margin-bottom: spacing();
  }

  &__background {
    background-color: color('grey', 'lighter');
    border-radius: border-radius('small');
    // provides some accessibility/contrast
    box-shadow: inset 0 0 1px color('night', 'light');
    height: 0.5em;
    margin-bottom: spacing('half');
    position: relative;

    &--small {
      height: 0.5em;
    }

    &--medium {
      height: 1em;
    }

    &--large {
      height: 2em;
    }
  }

  &__progress {
    height: 0;
    opacity: 0;
    overflow: hidden;
    width: 0;
  }

  &__indicator {
    background-color: color('green', 'light');
    border-radius: inherit;
    height: 100%;
    left: 0;
    position: absolute;
    transition: width duration('fast') easing('out');
    top: 0;

    @each $color-name in $button-colors {
      &--color-#{$color-name} {
        background-color: color($color-name);
      }
    }
  }

  &__message {
    text-align: center;
  }
}
