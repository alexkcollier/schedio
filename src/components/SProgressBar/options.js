export { colorNames as colors } from '@/utils/component-options';

export const sizes = [
  'small',
  'medium',
  'large',
];
