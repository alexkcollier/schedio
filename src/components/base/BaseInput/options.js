export const allowed = ['email', 'text', 'password', 'url', 'search', ''].sort()
