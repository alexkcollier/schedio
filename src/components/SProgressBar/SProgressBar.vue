<template>
  <div class="progress-bar">
    <div
      :class="progressBarClassList"
      class="progress-bar__background"
    >
      <progress
        :value="computedProgress"
        class="progress-bar__progress"
        max="100"
      />
      <div
        :class="progressIndicatorClassList"
        :style="{ width: `${computedProgress}%` }"
        class="progress-bar__indicator"
      />
    </div>
    <div
      v-if="computedMessage"
      class="progress-bar__message"
    >
      {{ computedMessage }}
    </div>
  </div>
</template>

<script>
import { colors, sizes } from './options';
import defaults from '@/utils/defaults';

const propError = (propName, values) => console.error(`\`${propName}\` must be one of ${values}`);

export default {
  name: 'SProgressBar',

  props: {
    progress: {
      type: [Number, String],
      required: true,
      validator (value) {
        if (value < 0) return console.error('Progress should be at least 0.');
        if (value > 100) return console.error('Progress should be less than 100.');
        return true;
      },
    },

    size: {
      type: String,
      default: '',
      vaidator: s => !s || sizes.includes(s) || propError('size', sizes),
    },

    color: {
      type: String,
      default: () => defaults.progressBarColor,
      vaidator: c => !c || colors.includes(c) || propError('color', colors),
    },

    message: {
      type: String,
      default: () => defaults.progressBarMessage,
    },

    hidePercentComplete: {
      type: Boolean,
      default: false,
    },
  },

  computed: {
    computedProgress () {
      if (this.progress < 0) return 0;
      if (this.progress > 100) return 100;
      return this.progress;
    },

    computedMessage () {
      let stack = this.message;
      if (this.message && !this.hidePercentComplete) stack += ': ';
      if (!this.hidePercentComplete) stack += `${this.computedProgress}% complete.`;

      return stack;
    },

    progressBarClassList () {
      return { [`progress-bar__background--${this.size}`]: this.size };
    },

    progressIndicatorClassList () {
      return { [`progress-bar__indicator--color-${this.color}`]: this.color };
    },
  },
};
</script>

<style lang="scss">
.progress-bar {
  margin-top: spacing();

  &:not(:last-child) {
    margin-bottom: spacing();
  }

  &__background {
    background-color: color('grey', 'lighter');
    border-radius: border-radius('small');
    // provides some accessibility/contrast
    box-shadow: inset 0 0 1px color('night', 'light');
    height: spacing('half');
    margin-bottom: spacing('half');
    position: relative;

    &--small {
      height: spacing('half');
    }

    &--medium {
      height: spacing();
    }

    &--large {
      height: spacing('double');
    }
  }

  &__progress {
    height: 0;
    opacity: 0;
    overflow: hidden;
    width: 0;
  }

  &__indicator {
    background-color: color('green', 'light');
    border-radius: inherit;
    height: 100%;
    left: 0;
    position: absolute;
    top: 0;
    transition: width duration('fast') easing('ease-out');

    @each $color-name in $button-colors {
      &--color-#{$color-name} {
        background-color: color($color-name);
      }
    }
  }

  &__message {
    text-align: center;
  }
}
</style>
