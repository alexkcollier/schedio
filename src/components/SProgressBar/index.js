import SProgressBar from './SProgressBar.vue';

export { SProgressBar };

export default {
  install: Vue => Vue.component(SProgressBar.name, SProgressBar),
};
