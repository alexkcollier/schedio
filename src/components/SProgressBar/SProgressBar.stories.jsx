import { SProgressBar } from '@/components/SProgressBar'
import { colors, sizes } from '@/components/SProgressBar/options'
import { SHeading } from '@/components/SHeading'
import PropList from '@@/docs/components/PropList'
import { boolean, number, select, text } from '@storybook/addon-knobs'
import { storiesOf } from '@storybook/vue'
import { generateHeading } from '@/utils/stories/render-functions'
import { SButton } from '@/components/SButton'

storiesOf('Components/Loading.SProgressBar', module)
  .addParameters({ jest: 'SProgressBar' })
  .add('Progress Bar', () => {
    return {
      props: {
        props: {
          default: () => ({
            progress: number('progress', 30, { range: true, min: 0, max: 100 }, 'Required Props'),
            size: select('size', ['', ...sizes], '', 'Optional Props'),
            color: select('color', ['', ...colors], '', 'Optional Props'),
            message: text('message', 'Something is happening', 'Optional Props'),
            hidePercentComplete: boolean('hide-percent-complete', true, 'Optional Props'),
          }),
        },
      },
      render (h) {
        return (
          <div>
            <SHeading level="1">Progress Bar</SHeading>
            <p>Progress bars visually represent the completion of a task.</p>

            <SHeading level="2">Example</SHeading>
            <SProgressBar {...{ props: this.$props.props }} />

            {SProgressBar.props && <PropList component={SProgressBar} />}
          </div>
        )
      },
    }
  })
  .add(
    'Progress Bar Examples',
    () => ({
      components: {
        AnimatedProgressBar: {
          name: 'AnimatedProgressBar',
          data () {
            return {
              start: null,
              barPosition: 0,
              shouldCountUp: true,
              duration: 10 * 1000, // 10 seconds
              req: null,
            }
          },
          destroyed () {
            this.cancelAnimation()
          },
          methods: {
            animateProgress (now) {
              /**
               * animate the progress bar over 10 seconds
               * Example: 10s = 10000ms, 10000ms / 100 ticks = 100 ms/tick
               */
              if (!this.start || now - this.start >= this.duration / 100) {
                this.start = now
                if (this.barPosition === 0) this.shouldCountUp = true

                if (this.barPosition === 100) this.shouldCountUp = false

                this.shouldCountUp ? this.barPosition++ : this.barPosition--
              }

              this.req = window.requestAnimationFrame(this.animateProgress)
            },
            cancelAnimation () {
              window.cancelAnimationFrame(this.req)
              this.req = null
            },
            startAnimation () {
              this.req = window.requestAnimationFrame(this.animateProgress)
            },
          },
          render (h) {
            return (
              <div>
                <SProgressBar progress={this.barPosition} message="Animation" />
                <SButton
                  color="night"
                  size="small"
                  is-outlined
                  onClick={() => (this.req ? this.cancelAnimation() : this.startAnimation())}
                >
                  {this.req ? 'Stop ' : 'Start '}
                  animation
                </SButton>
              </div>
            )
          },
        },
      },
      render (h) {
        return (
          <div>
            <SHeading>Progress Bar Examples</SHeading>

            <SHeading level="2">Animated</SHeading>
            <AnimatedProgressBar />

            <SHeading level="2">Colors</SHeading>
            {colors.map(color => generateProgressBar(h, color))}
          </div>
        )
      },
    }),
    {
      options: {
        showPanel: false,
      },
    }
  )

function generateProgressBar (h, color) {
  return [
    generateHeading(h, { level: 3, content: color }),
    h(SProgressBar, { props: { color, progress: Math.floor(Math.random() * 100) } }),
  ]
}
