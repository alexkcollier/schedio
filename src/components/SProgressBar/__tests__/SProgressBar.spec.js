import { shallowMount } from '@vue/test-utils'
import SProgressBar from '@/components/SProgressBar/SProgressBar.vue'
import * as options from '@/components/SProgressBar/options'

describe('SProgressBar.vue', () => {
  let wrapper
  const errorSpy = jest.spyOn(global.console, 'error').mockImplementation(() => {})

  beforeEach(() => {
    wrapper = shallowMount(SProgressBar, {
      propsData: {
        progress: '50'
      }
    })
  })

  afterEach(() => {
    jest.clearAllMocks()
  })

  it('renders correctly', () => {
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('renders progress', () => {
    shallowMount(SProgressBar)
    expect(errorSpy).toBeCalled()
  })

  it('only allows values between 0 and 100', () => {
    // we need to mount new components to test the prop validator
    const smallWrapper = shallowMount(SProgressBar, {
      propsData: {
        progress: -100
      }
    })
    const bigWrapper = shallowMount(SProgressBar, {
      propsData: {
        progress: 200
      }
    })

    wrapper.setProps({ progress: -100 })
    expect(smallWrapper.vm.computedProgress).toBe(0)
    wrapper.setProps({ progress: 200 })
    expect(bigWrapper.vm.computedProgress).toBe(100)
    // 4 times because Vue console.errors if prop is invalid, then we return consoe.error
    expect(errorSpy).toBeCalledTimes(4)
  })

  options.colors.forEach(color => {
    it(`can be ${color}`, () => {
      wrapper.setProps({ color })
      expect(errorSpy).not.toBeCalled()
      expect(wrapper.props('color')).toBe(color)
      expect(wrapper.html()).toMatchSnapshot()
    })
  })

  it('only allows valid colors', () => {
    shallowMount(SProgressBar, {
      propsData: {
        color: 'this is not a color'
      }
    })
    expect(errorSpy).toBeCalled()
  })

  options.sizes.forEach(size => {
    it(`can be ${size}`, () => {
      wrapper.setProps({ size })
      expect(errorSpy).not.toBeCalled()
      expect(wrapper.props('size')).toBe(size)
      expect(wrapper.html()).toMatchSnapshot()
    })
  })

  it('only allows valid sizes', () => {
    shallowMount(SProgressBar, {
      propsData: {
        size: 'this is not a size'
      }
    })
    expect(errorSpy).toBeCalled()
  })

  it('can display a message', () => {
    const message = 'this is a message'
    wrapper.setProps({ message })
    expect(wrapper.text()).toContain(message)
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('can hide the percent complete', () => {
    wrapper.setProps({ showPercentComplete: false })
    expect(wrapper.html()).toMatchSnapshot()
    expect(wrapper.text()).not.toContain(wrapper.props('progress') + '%')
  })
})
