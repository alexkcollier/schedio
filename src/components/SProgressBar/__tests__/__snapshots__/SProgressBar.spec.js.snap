// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SProgressBar.vue can be gold 1`] = `
<div class="progress-bar">
  <div class="progress-bar__background"><progress max="100" class="progress-bar__progress"></progress>
    <div class="progress-bar__indicator progress-bar__indicator--color-gold" style="width: 50%;"></div>
  </div>
  <div class="progress-bar__message">
    50% complete.
  </div>
</div>
`;

exports[`SProgressBar.vue can be green 1`] = `
<div class="progress-bar">
  <div class="progress-bar__background"><progress max="100" class="progress-bar__progress"></progress>
    <div class="progress-bar__indicator progress-bar__indicator--color-green" style="width: 50%;"></div>
  </div>
  <div class="progress-bar__message">
    50% complete.
  </div>
</div>
`;

exports[`SProgressBar.vue can be grey 1`] = `
<div class="progress-bar">
  <div class="progress-bar__background"><progress max="100" class="progress-bar__progress"></progress>
    <div class="progress-bar__indicator progress-bar__indicator--color-grey" style="width: 50%;"></div>
  </div>
  <div class="progress-bar__message">
    50% complete.
  </div>
</div>
`;

exports[`SProgressBar.vue can be ice 1`] = `
<div class="progress-bar">
  <div class="progress-bar__background"><progress max="100" class="progress-bar__progress"></progress>
    <div class="progress-bar__indicator progress-bar__indicator--color-ice" style="width: 50%;"></div>
  </div>
  <div class="progress-bar__message">
    50% complete.
  </div>
</div>
`;

exports[`SProgressBar.vue can be large 1`] = `
<div class="progress-bar">
  <div class="progress-bar__background progress-bar__background--large"><progress max="100" class="progress-bar__progress"></progress>
    <div class="progress-bar__indicator" style="width: 50%;"></div>
  </div>
  <div class="progress-bar__message">
    50% complete.
  </div>
</div>
`;

exports[`SProgressBar.vue can be magenta 1`] = `
<div class="progress-bar">
  <div class="progress-bar__background"><progress max="100" class="progress-bar__progress"></progress>
    <div class="progress-bar__indicator progress-bar__indicator--color-magenta" style="width: 50%;"></div>
  </div>
  <div class="progress-bar__message">
    50% complete.
  </div>
</div>
`;

exports[`SProgressBar.vue can be medium 1`] = `
<div class="progress-bar">
  <div class="progress-bar__background progress-bar__background--medium"><progress max="100" class="progress-bar__progress"></progress>
    <div class="progress-bar__indicator" style="width: 50%;"></div>
  </div>
  <div class="progress-bar__message">
    50% complete.
  </div>
</div>
`;

exports[`SProgressBar.vue can be night 1`] = `
<div class="progress-bar">
  <div class="progress-bar__background"><progress max="100" class="progress-bar__progress"></progress>
    <div class="progress-bar__indicator progress-bar__indicator--color-night" style="width: 50%;"></div>
  </div>
  <div class="progress-bar__message">
    50% complete.
  </div>
</div>
`;

exports[`SProgressBar.vue can be red 1`] = `
<div class="progress-bar">
  <div class="progress-bar__background"><progress max="100" class="progress-bar__progress"></progress>
    <div class="progress-bar__indicator progress-bar__indicator--color-red" style="width: 50%;"></div>
  </div>
  <div class="progress-bar__message">
    50% complete.
  </div>
</div>
`;

exports[`SProgressBar.vue can be small 1`] = `
<div class="progress-bar">
  <div class="progress-bar__background progress-bar__background--small"><progress max="100" class="progress-bar__progress"></progress>
    <div class="progress-bar__indicator" style="width: 50%;"></div>
  </div>
  <div class="progress-bar__message">
    50% complete.
  </div>
</div>
`;

exports[`SProgressBar.vue can be spartan_blue 1`] = `
<div class="progress-bar">
  <div class="progress-bar__background"><progress max="100" class="progress-bar__progress"></progress>
    <div class="progress-bar__indicator progress-bar__indicator--color-spartan_blue" style="width: 50%;"></div>
  </div>
  <div class="progress-bar__message">
    50% complete.
  </div>
</div>
`;

exports[`SProgressBar.vue can be turquoise 1`] = `
<div class="progress-bar">
  <div class="progress-bar__background"><progress max="100" class="progress-bar__progress"></progress>
    <div class="progress-bar__indicator progress-bar__indicator--color-turquoise" style="width: 50%;"></div>
  </div>
  <div class="progress-bar__message">
    50% complete.
  </div>
</div>
`;

exports[`SProgressBar.vue can be white 1`] = `
<div class="progress-bar">
  <div class="progress-bar__background"><progress max="100" class="progress-bar__progress"></progress>
    <div class="progress-bar__indicator progress-bar__indicator--color-white" style="width: 50%;"></div>
  </div>
  <div class="progress-bar__message">
    50% complete.
  </div>
</div>
`;

exports[`SProgressBar.vue can display a message 1`] = `
<div class="progress-bar">
  <div class="progress-bar__background"><progress max="100" class="progress-bar__progress"></progress>
    <div class="progress-bar__indicator" style="width: 50%;"></div>
  </div>
  <div class="progress-bar__message">
    this is a message: 50% complete.
  </div>
</div>
`;

exports[`SProgressBar.vue can hide the percent complete 1`] = `
<div class="progress-bar">
  <div class="progress-bar__background"><progress max="100" class="progress-bar__progress"></progress>
    <div class="progress-bar__indicator" style="width: 50%;"></div>
  </div>
  <!---->
</div>
`;

exports[`SProgressBar.vue renders correctly 1`] = `
<div class="progress-bar">
  <div class="progress-bar__background"><progress max="100" class="progress-bar__progress"></progress>
    <div class="progress-bar__indicator" style="width: 50%;"></div>
  </div>
  <div class="progress-bar__message">
    50% complete.
  </div>
</div>
`;
